### ISSUE TITLE
🐛 Bug: Recursive traversal causes OSError in most of dataset implementations

### ISSUE THREAD

**Issue by mthrok** [pytorch/audio#828]
## 🐛 Bug

As noted in https://github.com/pytorch/audio/issues/815, `walk_files` with `prefix=False` traverses directories recursively but only returns filenames. Therefore, having audio files in nested directory in the dataset directory causes Dataset implementation to fail. This affects VCTK, LibriSpeech, YesNo, LibriTTS.

## To Reproduce

Run the following script.

```python
import os
import shutil

from torchaudio.datasets.yesno import YESNO

root_dir = 'data'
dataset_dir = os.path.join(root_dir, 'waves_yesno')

dataset = YESNO(root_dir, download=True)
i = 0
for i, sample in enumerate(dataset):
    pass
print(f'iterated {i} items')

# Say "0_1_0_1_0_1_0_1.wav" is a custom data I created and I put it under "custom" dir.
custom_dir = os.path.join(dataset_dir, 'custom')
os.makedirs(custom_dir, exist_ok=True)
src = os.path.join(dataset_dir, '1_1_1_1_1_1_1_1.wav')
dst = os.path.join(custom_dir, '0_1_0_1_0_1_0_1.wav')
shutil.copyfile(src, dst)

print('created', dst)

dataset = YESNO(root_dir)
for sample in dataset:
    pass
print(f'iterated {i} items')
```

```
$ python foo.py
iterated 59 items
created data/waves_yesno/custom/0_1_0_1_0_1_0_1.wav
Traceback (most recent call last):
  File "foo.py", line 24, in <module>
    for sample in dataset:
  File "/scratch/moto/torchaudio/torchaudio/datasets/yesno.py", line 82, in __getitem__
    item = load_yesno_item(fileid, self._path, self._ext_audio)
  File "/scratch/moto/torchaudio/torchaudio/datasets/yesno.py", line 28, in load_yesno_item
    waveform, sample_rate = torchaudio.load(file_audio)
  File "/scratch/moto/torchaudio/torchaudio/backend/sox_backend.py", line 35, in load
    raise OSError("{} not found or is a directory".format(filepath))
OSError: data/waves_yesno/0_1_0_1_0_1_0_1.wav not found or is a directory
```

## Expected behavior

It should not fail because of another wav file exists.

Dataset implementations should only care about handling the official dataset correctly. If one has to use custom dataset, they should be using [ConcatDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.ConcatDataset) and such.

None of the torchaudio dataset requires the functionality of `walk_files` to recursively traverse directories. Therefore we should remove it.

**Comment by vincentqb** [comment]
> ## 🐛 Bug
> As noted in #815, `walk_files` with `prefix=False` traverses directories recursively but only returns filenames. Therefore, having audio files in nested directory in the dataset directory causes Dataset implementation to fail. This affects VCTK, LibriSpeech, YesNo, LibriTTS.
> 
> ## To Reproduce
> Run the following script.
> 
> ```python
> import os
> import shutil
> 
> from torchaudio.datasets.yesno import YESNO
> 
> root_dir = 'data'
> dataset_dir = os.path.join(root_dir, 'waves_yesno')
> 
> dataset = YESNO(root_dir, download=True)
> i = 0
> for i, sample in enumerate(dataset):
>     pass
> print(f'iterated {i} items')
> 
> # Say "0_1_0_1_0_1_0_1.wav" is a custom data I created and I put it under "custom" dir.
> custom_dir = os.path.join(dataset_dir, 'custom')
> os.makedirs(custom_dir, exist_ok=True)
> src = os.path.join(dataset_dir, '1_1_1_1_1_1_1_1.wav')
> dst = os.path.join(custom_dir, '0_1_0_1_0_1_0_1.wav')
> shutil.copyfile(src, dst)
> 
> print('created', dst)
> 
> dataset = YESNO(root_dir)
> for sample in dataset:
>     pass
> print(f'iterated {i} items')
> ```
> 
> ```
> $ python foo.py
> iterated 59 items
> created data/waves_yesno/custom/0_1_0_1_0_1_0_1.wav
> Traceback (most recent call last):
>   File "foo.py", line 24, in <module>
>     for sample in dataset:
>   File "/scratch/moto/torchaudio/torchaudio/datasets/yesno.py", line 82, in __getitem__
>     item = load_yesno_item(fileid, self._path, self._ext_audio)
>   File "/scratch/moto/torchaudio/torchaudio/datasets/yesno.py", line 28, in load_yesno_item
>     waveform, sample_rate = torchaudio.load(file_audio)
>   File "/scratch/moto/torchaudio/torchaudio/backend/sox_backend.py", line 35, in load
>     raise OSError("{} not found or is a directory".format(filepath))
> OSError: data/waves_yesno/0_1_0_1_0_1_0_1.wav not found or is a directory
> ```

This seems to be due to how the path is regenerated when accessing a file, right?
> ## Expected behavior
> It should not fail because of another wav file exists.

To make this more precise, my impression would be that the dataset should silently accommodate modified datasets, and we should define how the abstraction would behave when presented with modifications.

> Dataset implementations should only care about handling the official dataset correctly. If one has to use custom dataset, they should be using [ConcatDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.ConcatDataset) and such.

I'm expecting users to do modifications to their datasets. Based this comment, are you advocating that the dataset should silently omit files that have been added? How about files that are removed?

> None of the torchaudio dataset requires the functionality of `walk_files` to recursively traverse directories. Therefore we should remove it.

To clarify, are you saying no dataset uses recursion? or that defining `walk_files` is not needed as a new function?

**Comment by mthrok** [comment]
Resolved by #1109
